Modules/Nonunit/Review/include/itkAttributeMorphologyBaseImageFilter.hxx

```diff
@@ -198,10 +199,10 @@ AttributeMorphologyBaseImageFilter<TInputImage, TOutputImage, TAttribute, TFunct
     progress.CompletedPixel();
   }
 
-  delete[] m_Raw;
-  delete[] m_SortPixels;
-  delete[] m_Parent;
-  delete[] m_AuxData;
+  m_Raw.reset();
+  m_SortPixels.reset();
+  m_Parent.reset();
+  m_AuxData.reset();
```

Comment:
Note that these four `reset()` calls would not have been necessary if m_Raw, m_SortPixels, m_Parent, and m_AuxData would have been local variables, declared within the member function itself.

Reply:
@N-Dekker I think that the better way to handle this problem is to convert the instance variables into local variables for the GenerateData method.

It may require also converting some of the classes' private member functions into lambda functions inside of GenerateData.  (i.e. Criterion, Union )

NOTE:  Your simple .reset solution is an improvement, so I'd be happy to add it as an evolutionary step forward.

Reply:
> I think that the better way to handle this problem is to convert the instance variables into local variables for the GenerateData method.

@hjmjohnson I agree, but converting them into local variables doesn't appear entirely trivial. Some private member functions of AttributeMorphologyBaseImageFilter depend on them being  instance variables. While I believe those private member functions could be adjusted (probably converted to lambda's), I feel reluctant to do so, as I'm not familiar with this filter, and I'm not using it either.

Reply:
For that next pull request, can we get a member`.reset()` call invoked as part of the desctructor of a local variable of this `GenerateData` method?  That way, we wouldn't have to worry about what path is taken in exiting this method.

Reply:
> That way, we wouldn't have to worry about what path is taken in exiting this method.

@Leengit Thanks for your feedback. Honestly, I don't worry too much, because if the taken path does _not_ reach the four proposed `reset()` calls (typically because of an exception somewhere inside GenerateData), the memory is still owned by those four unique_ptrs (m_Raw, m_SortPixels, m_Parent, and m_AuxData), so it won't leak. The memory will then be freed by the next GenerateData call on that filter, or by the destructor of the filter.

Reply:
Agreed that the current pull request is a significant step in the right direction and is plenty good as is.

Giving some additional information for my idea in case we do revisit this ... I looked for a standard C++ template that would accept an arbitrary method rather than hardcoding `reset`, but I couldn't find one.  That we might have to code things ourselves makes things uglier; we might have to create something like
```c++
template <typename T> class ResetOnDestruction {
private:
  T &m_resetable;

public:
  ResetOnDestruction(ResetOnDestruction &) = delete;
  ResetOnDestruction(T &resetable) : m_resetable(resetable) {}
  ~ResetOnDestruction() { m_resetable.reset(); }
};
```
... which would then be used as in:
```c++
ResetOnDesctruction<decltype(m_SortPixels)> gc_SortPixels(m_SortPixels);
m_SortPixels = make_unique_for_overwrite<OffsetValueType[]>(buffsize);
```
... so that `m_SortPixels.reset()` is invoked when `gc_SortPixels` goes out of scope.

Reply:
> freed by the next GenerateData call on that filter

Indeed, much less terrible than a typical leak!  Still "if everyone did this", we could have a bunch of ITK objects holding onto extra memory they don't need at the moment.  This is not the highest priority in the overall scheme of things, but I'd venture that it's worth fixing if it catches someone's interest.

Reply:
I understand your point, @Leengit but I still think that the _ideal_ solution would be to declare those four unique_ptrs as local variables, instead of as member variables. Which would automatically cleanup the memory when exiting GenerateData, on each possible path.

Reply:
> I looked for a standard C++ template that would accept an arbitrary method rather than hardcoding reset, but I couldn't find one.

Could it be that you were looking for `unique_resource`?  https://en.cppreference.com/w/cpp/experimental/unique_resource